SonarQube.Client/Models/ServerSentEvents/SSEStreamWriter.cs

```diff
@@ -35,39 +37,60 @@ public interface ISSEStreamWriter : IDisposable
         Task BeginListening();
     }
 
+    /// <summary>
+    /// Aggregates stream lines into events.
+    /// Code on the java side: https://github.com/SonarSource/sonarlint-core/blob/171ca4d75c24033e115a81bd7481427cd1f39f4c/server-api/src/main/java/org/sonarsource/sonarlint/core/serverapi/stream/EventBuffer.java
+    /// </summary>
     internal class SSEStreamWriter : ISSEStreamWriter
     {
-        private readonly Stream serverStream;
+        private readonly StreamReader serverStreamReader;
         private readonly ChannelWriter<ISqServerEvent> sqEventsChannel;
         private readonly CancellationToken cancellationToken;
         private readonly ISqServerSentEventParser sqServerSentEventParser;
 
-        public SSEStreamWriter(Stream serverStream,
+        public SSEStreamWriter(StreamReader serverStreamReader,
             ChannelWriter<ISqServerEvent> sqEventsChannel,
             CancellationToken cancellationToken)
-            : this(serverStream, sqEventsChannel, cancellationToken, new SqServerSentEventParser())
+            : this(serverStreamReader, sqEventsChannel, cancellationToken, new SqServerSentEventParser())
         {
         }
 
-        internal SSEStreamWriter(Stream serverStream,
+        internal SSEStreamWriter(StreamReader serverStreamReader,
             ChannelWriter<ISqServerEvent> sqEventsChannel,
             CancellationToken cancellationToken,
             ISqServerSentEventParser sqServerSentEventParser)
         {
-            this.serverStream = serverStream;
+            this.serverStreamReader = serverStreamReader;
             this.cancellationToken = cancellationToken;
             this.sqEventsChannel = sqEventsChannel;
             this.sqServerSentEventParser = sqServerSentEventParser;
         }
 
-        public Task BeginListening()
+        public async Task BeginListening()
         {
-            // todo: next PR
-            throw new NotImplementedException();
+            var eventLines = new List<string>();
+
+            while (!serverStreamReader.EndOfStream && !cancellationToken.IsCancellationRequested)
+            {
+                var line = await serverStreamReader.ReadLineAsync();
+                var isEventEnd = line == string.Empty;
+
+                if (isEventEnd)
+                {
+                    var parsedEvent = sqServerSentEventParser.Parse(eventLines.ToList());
```

Comment:
I don't understand why, but if I don't pass `eventLines.ToList()`, sqServerSentEventParser's mock invocations fail that it expected the wrong thing.

Reply:
Allocating an extra list is not good. Fixing or even removing the test is better

Reply:
For some reason the test passes now. Hmm.